### PR TITLE
`Development`: Refactor quiz exercise service

### PR DESCRIPTION
**Pull request by rriyaldhi**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context

The implementation of Quiz Pool API will reuse the functionalities of Quiz Exercise API. Therefore, QuizExerciseService is refactored so that the functionality can be reused later by Quiz Pool.

### Description

1. Create QuizService class
QuizService is an abstract class that contains the common methods of QuizExerciseService and QuizPoolService (will be implemented on separate PR).
2. Move the save, saveCorrectMappingsInIndices, saveCorrectMappingsInIndicesShortAnswer, restoreCorrectMappingsFromIndices, and restoreCorrectMappingsFromIndicesShortAnswer methods to QuizService.
The methods, which are used by QuizExerciseService, will be used by the QuizPoolService as well.
3. Create QuizConfiguration Interface
QuizConfiguration is an abstraction that contains the common operation of QuizExercise and QuizPool (will be implemented on separate PR)
4. Move the reconnectJSONIgnoreAttributes method to QuizConfiguration
The reconnectJSONIgnoreAttributes method, which is used by QuizExercise, will be used by the QuizPool as well.
5. Refactor QuizConfiguration.reconnectJSONIgnoreAttributes and QuizService.save method

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course

1. Log in to Artemis
2. Navigate to Course Administration
3. Select the Course
7. Navigate to Exercises
8. Create a new Quiz Exercise
9. Verify that the Quiz Exercise can be created successfully.
10. Edit the newly created Quiz Exercise
11. Verify that the Quiz Exercise can be updated successfully.
12. Verify that the Quiz Exercise can be reevaluated successfully.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| QuizConfiguration.java | 78% | 100% |
| QuizExercise.java | 73% | 91% |
| QuizExerciseService.java | 92% | 100% |
| QuizService.java | 82% | 94% |
| AnswerCounter.java | n/a | 100% |
| AnswerOption.java | 50% | 94% |
| DragAndDropMapping.java | 50% | 90% |
| DragAndDropQuestion.java | 49% | 83% |
| DragItem.java | 33% | 82% |
| DropLocation.java | 58% | 94% |
| DropLocationCounter.java | n/a | 100% |
| MultipleChoiceQuestion.java | 62% | 88% |
| QuizQuestion.java | 66% | 96% |
| QuizQuestionStatistic.java | n/a | 92% |
| ShortAnswerMapping.java | 75% | 97% |
| ShortAnswerQuestion.java | 55% | 87% |
| ShortAnswerSolution.java | 50% | 93% |
| ShortAnswerSpot.java | 75% | 97% |
| ShortAnswerSpotCounter.java | n/a | 100% |

### Screenshots

https://user-images.githubusercontent.com/10885503/230736567-c00948a3-d60f-4a82-af8a-fa79f23c7643.mov

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+     * Set the parent of the given QuizQuestion to reconnect ignored JSON attributes.
+     *
+     * @param quizQuestion the QuizQuestion of which the parent to be set
+     */
+    void setQuestionParent(QuizQuestion quizQuestion);
+
+    /**
+     * Recreate missing pointers from children to parents that were removed by @JSONIgnore
+     */
+    default void reconnectJSONIgnoreAttributes() {// iterate through quizQuestions to add missing pointer back to quizExercise
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java:24]: I know you just moved this code, but I think this is a good time to refactor this function. Please split up logic for the different instances into their own functions (potentially private static functions).

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                            dragItem.setQuestion(dragAndDropQuestion);
+                        }
+                    }
+                    // reconnect correctMappings
+                    for (DragAndDropMapping mapping : dragAndDropQuestion.getCorrectMappings()) {
+                        if (mapping.getId() != null) {
+                            mapping.setQuestion(dragAndDropQuestion);
+                        }
+                    }
+                    // reconnect dropLocationCounters
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java:70]: This is three times the same for loop. You could generalize this for loop using a function with a generic function parameter.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                            solution.setQuestion(shortAnswerQuestion);
+                        }
+                    }
+                    // reconnect correctMappings
+                    for (ShortAnswerMapping mapping : shortAnswerQuestion.getCorrectMappings()) {
+                        if (mapping.getId() != null) {
+                            mapping.setQuestion(shortAnswerQuestion);
+                        }
+                    }
+                    // reconnect spotCounters
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java:98]: This is the same for loop again. Also generalize this please.

File: src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java [hunk 4]
```diff
@@ -480,4 +201,50 @@ private void resetInvalidQuestions(QuizExercise quizExercise) {
...
+            var quizPointStatistic = new QuizPointStatistic();
+            quizExercise.setQuizPointStatistic(quizPointStatistic);
+            quizPointStatistic.setQuiz(quizExercise);
+        }
+
+        // make sure the pointers in the statistics are correct
+        quizExercise.recalculatePointCounters();
+
+        QuizExercise savedQuizExercise = super.save(quizExercise);
+
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java:220]: Within a function, please do not mix var variable declarations with traditional variable declarations. Please stay consistent.

File: src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java [hunk 5]
```diff
@@ -480,4 +201,50 @@ private void resetInvalidQuestions(QuizExercise quizExercise) {
...
+    protected QuizExercise saveAndFlush(QuizExercise quizExercise) {
+        if (quizExercise.getQuizBatches() != null) {
+            for (QuizBatch quizBatch : quizExercise.getQuizBatches()) {
+                quizBatch.setQuizExercise(quizExercise);
+                if (quizExercise.getQuizMode() == QuizMode.SYNCHRONIZED) {
+                    if (quizBatch.getStartTime() != null) {
+                        quizExercise.setDueDate(quizBatch.getStartTime().plusSeconds(quizExercise.getDuration() + Constants.QUIZ_GRACE_PERIOD_IN_SECONDS));
+                    }
+                }
+                else {
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java:239]: Is this intended indentation?

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+            }
+            else if (quizQuestion instanceof ShortAnswerQuestion shortAnswerQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndicesShortAnswer(shortAnswerQuestion);
+            }
+        }
+
+        return savedQuizConfiguration;
+    }
+
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizService.java:140]: Please stay consistent with your variable declarations and refactor each instanceof case into its own function. The for loops in line 57 and 81 and 109 are the same code, you can refactor it into a function with a generic parameter.

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            boolean dropLocationFound = false;
+            for (DropLocation questionDropLocation : dragAndDropQuestion.getDropLocations()) {
+                if (dropLocation.equals(questionDropLocation)) {
+                    dropLocationFound = true;
+                    mapping.setDropLocationIndex(dragAndDropQuestion.getDropLocations().indexOf(questionDropLocation));
+                    mapping.setDropLocation(null);
+                    break;
+                }
+            }
+
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizService.java:178]: Code duplication, refactor into a function with generic parameter and return a boolean if the item has been found.

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+            boolean spotFound = false;
+            for (ShortAnswerSpot questionSpot : shortAnswerQuestion.getSpots()) {
+                if (spot.equals(questionSpot)) {
+                    spotFound = true;
+                    mapping.setShortAnswerSpotIndex(shortAnswerQuestion.getSpots().indexOf(questionSpot));
+                    mapping.setSpot(null);
+                    break;
+                }
+            }
+
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizService.java:227]: Code duplication. See other comments.

**Review by TimOrtel — CHANGES_REQUESTED** [review]
Thank you for this PR. I think its a good start but you can still improve on this code and clean up the duplicated code.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java (hunk 1 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java:25]: code is refactored

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java (hunk 2 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java]: loop is moved to a generalized method.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java (hunk 3 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java]: done

File: src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java (hunk 4 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java]: change the var variable

File: src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java (hunk 5 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizExerciseService.java:239]: fix the indentation

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java (hunk 6 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizService.java]: Codes are moved to separate functions for each case and generic method findTobeDeletedComponent is created

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java (hunk 8 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizService.java]: Code is refactored

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java (hunk 7 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizService.java]: Code is refactored

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS2 in the testing session, all seems to be working properly

**Review by danielbode — APPROVED** [review]
Tested on ts2 during a testing session. Still works as expected.

**Review by DominikRemo — APPROVED** [review]
Tested in testing session on ts2. Works as expected

**Review by chrisknedl — APPROVED** [review]
Tested on TS2 during today's testing session. Everything is working.

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+            else if (quizQuestion instanceof ShortAnswerQuestion shortAnswerQuestion) {
+                // restore references from index after save
+                restoreCorrectMappingsFromIndicesShortAnswer(shortAnswerQuestion);
+            }
+        }
+
+        return savedQuizConfiguration;
+    }
+
+    private void fixReferenceMultipleChoice(MultipleChoiceQuestion mcQuestion) {
```
Comment (TimOrtel) [inline src/main/java/de/tum/in/www1/artemis/service/QuizService.java:69]: This is the (almost identical) same function three times in a row.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java (hunk 1 above)
Comment (TimOrtel) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizConfiguration.java:25]: Nice!

**Review by TimOrtel — CHANGES_REQUESTED** [review]
The applies refactorings look good. I think there is one more refactoring that is needed and then we can merge this.

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestionStatisticComponent.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package de.tum.in.www1.artemis.domain.quiz;
+
+public interface QuizQuestionStatisticComponent<T1 extends QuizQuestionStatistic, T2 extends QuizQuestionComponent<T3>, T3 extends QuizQuestion> {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestionStatisticComponent.java:3]:
```suggestion
public interface QuizQuestionStatisticComponent<S extends QuizQuestionStatistic, C extends QuizQuestionComponent<T3>, Q extends QuizQuestion> {
```
I think the enumeration is unnecessary and makes it more complicated.

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        saveCorrectMappingsInIndicesShortAnswer(saQuestion);
+    }
+
+    /**
+     * Call the given deleteCallback for each given component where id is not null
+     *
+     * @param components     the components for which the given deleteCallback to be applied to
+     * @param deleteCallback the callback to be called for each component
+     */
+    private <T1 extends QuizQuestionStatisticComponent<T2, T3, T4>, T2 extends QuizQuestionStatistic, T3 extends QuizQuestionComponent<T4>, T4 extends QuizQuestion> void findTobeDeletedComponent(
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/QuizService.java:154]: I would also suggest resolving the enumeration here and below

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java (hunk 9 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizService.java]: refactored

File: src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestionStatisticComponent.java (hunk 10 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/domain/quiz/QuizQuestionStatisticComponent.java:3]: thanks

File: src/main/java/de/tum/in/www1/artemis/service/QuizService.java (hunk 11 above)
Comment (rriyaldhi) [reply src/main/java/de/tum/in/www1/artemis/service/QuizService.java]: enumeration is removed

**Review by TheZoker — DISMISSED** [review]
Tested during testing session on TS2 and everything worked as expected! 👍

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts2. Creating, editing and reevaluating a quiz worked as expected.

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested on TS2 and all seems to be working as expected, but wanted to comment something:

Once the exam starts, the edit screen does not automatically block all the edits, but updates after some time. Editing the details of the quiz does not change anything anyways from what I have tested.

Screenshot where top screen was updated manually.
Screenshot 2023-04-28 135306

**Review by pellumbbaboci — APPROVED** [review]
Tested on ts3 during testing session, works as expected 👍

**Review by YeetTheFirst21 — APPROVED** [review]
Checked again on TS1, did all the steps @Santia-go mentioned before as well. Plus checked file export/import and found no issues

**Pull request merged**